c/cert/src/rules/ENV30-C/DoNotModifyTheReturnValueOfCertainFunctions.ql

```diff
@@ -0,0 +1,68 @@
+/**
+ * @id c/cert/do-not-modify-the-return-value-of-certain-functions
+ * @name ENV30-C: Do not modify the return value of certain functions
+ * @description Return value of getenv and similar functions cannot be modified.
+ * @kind path-problem
+ * @precision very-high
+ * @problem.severity error
+ * @tags external/cert/id/env30-c
+ *       correctness
+ *       external/cert/obligation/rule
+ */
+
+import cpp
+import codingstandards.c.cert
+import semmle.code.cpp.dataflow.DataFlow
+import DataFlow::PathGraph
+
+/*
+ * Call to functions that return pointers to environment objects that should not be modified.
+ */
+
+class NotModifiableCall extends FunctionCall {
+  NotModifiableCall() {
+    this.getTarget()
+        .hasGlobalOrStdName(["getenv", "setlocale", "localeconv", "asctime", "strerror"])
+  }
+}
+
+/*
+ * An expression that modifies an object.
+ */
+
+class ObjectWrite extends Expr {
+  ObjectWrite() {
+    // the pointed object is reassigned
+    exists(Expr e |
+      e = [any(AssignExpr ae).getLValue(), any(CrementOperation co).getOperand()] and
+      (
+        this = e.(PointerDereferenceExpr).getOperand()
+        or
+        this = e.(PointerFieldAccess).getQualifier()
+      )
+    )
+  }
+}
+
+/**
+ * DF configuration for flows from a `NotModifiableCall` to a object modifications.
+ */
+class DFConf extends DataFlow::Configuration {
```

Comment:
how does this rule feel about modifying a member of a struct, for which a pointer to that struct is returned from one of those calls? does that count as a modification to the object overall? will DF be enough to see those, if this rule does care about that?

Reply:
oh , I do see that there is a testcase for this, I guess I am still unsure if this still requires consideration for the case with a plain pointer reassignment to the struct (as opposed to the other testcase with a distinct struct)

Reply:
actually this is fine in a way I did not expect/the aliasing is handled, but has an additional alert that I did not expect:
```
void f6(void) {
  struct lconv *conv4 = localeconv(); //this line is also reported?
  struct lconv *convAlias = conv4;
  
  if ('\0' == convAlias->decimal_point[0]) {
    convAlias->decimal_point = "."; // NON_COMPLIANT - accurately detected
  }
}
```

Reply:
oh, wait I totally misread that, it was an entry in the nodes of the path. Makes sense now!